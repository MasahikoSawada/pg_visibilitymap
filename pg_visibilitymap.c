/*-------------------------------------------------------------------------
 *
 * pg_visibilitymap.c
 *  display contents of a visibility map
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/visibilitymap.h"
#include "funcapi.h"
#include "storage/bufmgr.h"

PG_MODULE_MAGIC;

/*
 * Returns visibility on a given page, according to the visibility map.
 */
PG_FUNCTION_INFO_V1(pg_is_all_visible);

/*
 * Return the page is all-visible or not, according to the visibility map.
 */
Datum
pg_is_all_visible(PG_FUNCTION_ARGS)
{
	Oid		relid = PG_GETARG_OID(0);
	int64	blkno = PG_GETARG_INT64(1);
	Relation rel;
	Buffer	vmbuffer = InvalidBuffer;
	bool	result = false;

	rel = relation_open(relid, AccessShareLock);

	if (blkno < 0 || blkno > MaxBlockNumber)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("invalid block number")));

	result = visibilitymap_test(rel, blkno, &vmbuffer);

	if (BufferIsValid(vmbuffer))
		ReleaseBuffer(vmbuffer);
	relation_close(rel, AccessShareLock);

	PG_RETURN_BOOL(result);
}
