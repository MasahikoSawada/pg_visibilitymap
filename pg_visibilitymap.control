# pg_visibilitymapmap extension
comment = 'examine the VisibilityMap (VM)'
default_version = '1.0'
module_pathname = '$libdir/pg_visibilitymap'
relocatable = true