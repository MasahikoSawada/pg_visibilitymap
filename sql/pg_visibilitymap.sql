create extension pg_visibilitymap;

CREATE TABLE t1 AS SELECT * FROM generate_series(1,1000) AS p;
\d t1
VACUUM t1;
SELECT relpages FROM pg_class WHERE relname = 't1';
SELECT * FROM pg_visibilitymap('t1');
DELETE FROM t1 WHERE p < 500;
SELECT * FROM pg_visibilitymap('t1');
VACUUM t1;
SELECT * FROM pg_visibilitymap('t1');
DROP TABLE t1;

--
-- Errors
--
CREATE TABLE t2 (col int primary key);
INSERT INTO t2 SELECT generate_series(1,100);
CREATE VIEW t2_view AS (SELECT * FROM t2);
CREATE MATERIALIZED VIEW t2_matview AS (SELECT * FROM t2);
SELECT pg_is_all_visible('t2', 0);
SELECT pg_is_all_visible('t2_view', 0);
SELECT pg_is_all_visible('t2_matview', 0);
DROP TABLE t2 CASCADE;