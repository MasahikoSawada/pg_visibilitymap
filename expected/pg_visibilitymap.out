create extension pg_visibilitymap;
CREATE TABLE t1 AS SELECT * FROM generate_series(1,1000) AS p;
\d t1
      Table "public.t1"
 Column |  Type   | Modifiers 
--------+---------+-----------
 p      | integer | 

VACUUM t1;
SELECT relpages FROM pg_class WHERE relname = 't1';
 relpages 
----------
        5
(1 row)

SELECT * FROM pg_visibilitymap('t1');
 blkno | all_visible 
-------+-------------
     0 | t
     1 | t
     2 | t
     3 | t
     4 | t
(5 rows)

DELETE FROM t1 WHERE p < 500;
SELECT * FROM pg_visibilitymap('t1');
 blkno | all_visible 
-------+-------------
     0 | f
     1 | f
     2 | f
     3 | t
     4 | t
(5 rows)

VACUUM t1;
SELECT * FROM pg_visibilitymap('t1');
 blkno | all_visible 
-------+-------------
     0 | t
     1 | t
     2 | t
     3 | t
     4 | t
(5 rows)

DROP TABLE t1;
--
-- Errors
--
CREATE TABLE t2 (col int primary key);
INSERT INTO t2 SELECT generate_series(1,100);
CREATE VIEW t2_view AS (SELECT * FROM t2);
CREATE MATERIALIZED VIEW t2_matview AS (SELECT * FROM t2);
SELECT pg_is_all_visible('t2', 0);
 pg_is_all_visible 
-------------------
 f
(1 row)

SELECT pg_is_all_visible('t2_view', 0);
ERROR:  "t2_view" is not a table or materialized view
SELECT pg_is_all_visible('t2_matview', 0);
 pg_is_all_visible 
-------------------
 f
(1 row)

DROP TABLE t2 CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to view t2_view
drop cascades to materialized view t2_matview
